netann/channel_announcement.go

```diff
@@ -175,3 +196,96 @@ func validateChannelAnn1(a *lnwire.ChannelAnnouncement1) error {
 
 	return nil
 }
+
+// validateChannelAnn2 validates the channel announcement message and checks
+// that message signature covers the announcement message.
+func validateChannelAnn2(a *lnwire.ChannelAnnouncement2,
+	fetchPkScript FetchPkScript) error {
+
+	dataHash, err := ChanAnn2DigestToSign(a)
+	if err != nil {
+		return err
+	}
+
+	sig, err := a.Signature.ToSignature()
+	if err != nil {
+		return err
+	}
+
+	nodeKey1, err := btcec.ParsePubKey(a.NodeID1.Val[:])
+	if err != nil {
+		return err
+	}
+
+	nodeKey2, err := btcec.ParsePubKey(a.NodeID2.Val[:])
+	if err != nil {
+		return err
+	}
+
+	keys := []*btcec.PublicKey{
+		nodeKey1, nodeKey2,
+	}
+
+	// If the bitcoin keys are provided in the announcement, then it is
+	// assumed that the signature of the announcement is a 4-of-4 MuSig2
+	// over the bitcoin keys and node ID keys.
+	if a.BitcoinKey1.IsSome() && a.BitcoinKey2.IsSome() {
+		var (
+			btcKey1 tlv.RecordT[tlv.TlvType12, [33]byte]
+			btcKey2 tlv.RecordT[tlv.TlvType14, [33]byte]
+		)
+
+		btcKey1 = a.BitcoinKey1.UnwrapOr(btcKey1)
+		btcKey2 = a.BitcoinKey2.UnwrapOr(btcKey2)
+
+		bitcoinKey1, err := btcec.ParsePubKey(btcKey1.Val[:])
+		if err != nil {
+			return err
+		}
+
+		bitcoinKey2, err := btcec.ParsePubKey(btcKey2.Val[:])
+		if err != nil {
+			return err
+		}
+
+		keys = append(keys, bitcoinKey1, bitcoinKey2)
+	} else {
```

Comment:
Q: what if only one of the keys is set, like `bitcoinKey1` is set but not `bitcoinKey2`?

Reply:
it will then fall to the `else` branch below which will fetch the on-chain script. 
ie, we dont explicitly need to not allow that case I dont think. It could maybe be a valid future message potentially.